src/compiler/driver.cc

```diff
@@ -343,11 +328,31 @@ hilti::Result<hilti::Nothing> Driver::compile() {
     return hilti::Nothing();
 }
 
-hilti::Result<UnitInfo> Driver::lookupUnit(const hilti::ID& unit) {
-    if ( auto x = _units.find(unit); x != _units.end() )
+hilti::Result<TypeInfo> Driver::lookupType(const hilti::ID& id) {
+    if ( auto x = _types.find(id); x != _types.end() )
         return x->second;
     else
-        return hilti::result::Error("unknown unit");
+        return hilti::result::Error(hilti::util::fmt("unknown type '%s'", id));
+}
+
+std::vector<TypeInfo> Driver::types(bool exported_only) const {
+    auto exported = _glue->exportedIDs();
+
+    std::vector<TypeInfo> result;
+    result.reserve(_types.size());
+    for ( const auto& t : _types ) {
+        if ( exported_only && std::find_if(exported.begin(), exported.end(),
+                                           [&t](const auto& i) { return i.first == t.first; }) == exported.end() )
+            continue;
+
+        result.push_back(t.second);
+    }
```

Comment:
Worst case this will be a quadratic loop in the number of types.

For the `exported_only` case we could simply iterate over `exported` and lookup `TypeInfo`s in `_types`. We could add a branch doing that. In that case we should also adjust the `reserve` to only ask for `exported.size()`.

Reply:
I'll leave this as is, `exported_types` is going to remain tiny.

Reply:
Well, turns out I'm moving the `exported_only` cases into a separate functions, which then indeed ends up iterating over `exported` to ensure a consistent order.